src/Features/Core/Portable/CodeRefactorings/SyncNamespace/AbstractChangeNamespaceService.cs

```diff
@@ -102,6 +106,67 @@ public override async Task<bool> CanChangeNamespaceAsync(Document document, Synt
             return !applicableContainers.IsDefault;
         }
 
+        public override string TryBuildNamespaceFromFolders(IEnumerable<string> folders, ISyntaxFacts syntaxFacts)
+        {
+            var parts = folders.SelectMany(folder => folder.Split(new[] { '.' })).SelectAsArray(EscapeIdentifier);
+            return parts.All(syntaxFacts.IsValidIdentifier) ? string.Join(".", parts) : null;
+        }
+
+        public override async Task<Solution> ChangeTopLevelNamespacesAsync(
+            Document document,
+            string targetNamespace,
+            CancellationToken cancellationToken)
+        {
+            var syntaxFacts = document.GetLanguageService<ISyntaxFactsService>();
+            var syntaxRoot = await document.GetRequiredSyntaxRootAsync(cancellationToken).ConfigureAwait(false);
+
+            // Don't descend into anything other than top level declarations from the root.
+            // ChangeNamespaceService only controls top level declarations right now.
+            // Don't use namespaces that already match the target namespace
+            var namespaceDeclaration = syntaxRoot
+                .DescendantNodes(n => !syntaxFacts.IsDeclaration(n))
+                .FirstOrDefault(n => syntaxFacts.IsNamespaceDeclaration(n));
+
+            if (namespaceDeclaration is null)
+            {
+                throw new InvalidOperationException("Unable to find valid top level namespaces to change");
```

Comment:
if this is a public entrypoint, this needs to be a localized resource.

Reply:
I'm going to change this to TryChangeTopLevelNamespaceAsync and return null here instead